providers/desec/protocol.go

```diff
@@ -0,0 +1,215 @@
+	}
+	err = json.Unmarshal(bodyString, &dr)
+	if err != nil {
+		return err
+	}
+	for _, domain := range dr {
+		//We store the min ttl in the domain index
+		//This will be used for validation and auto correction
+		c.domainIndex[domain.Name] = domain.MinimumTTL
+	}
+	return nil
+}
+
+func (c *api) getRecords(domain string) ([]resourceRecord, error) {
+	endpoint := "/domains/%s/rrsets/"
+	var rrs []rrResponse
+	var rrs_new []resourceRecord
+	var bodyString, err = c.get(fmt.Sprintf(endpoint, domain), "GET")
+	if err != nil {
+		return rrs_new, fmt.Errorf("Error fetching records from deSEC for domain %s: %s", domain, err)
+	}
+	err = json.Unmarshal(bodyString, &rrs)
+	if err != nil {
+		return rrs_new, err
+	}
+	// deSEC returns round robin records as array but dnsconfig expects single entries for each record
+	// we will create one object per record except of TXT records which are handled as array of string by dnscontrol aswell.
+	for i := range rrs {
+		tmp := resourceRecord{
+			TTL:     rrs[i].TTL,
+			Type:    rrs[i].Type,
+			Subname: rrs[i].Subname,
+			Records: rrs[i].Records,
+		}
+		rrs_new = append(rrs_new, tmp)
+	}
+	return rrs_new, nil
+}
+
+func (c *api) createDomain(domain string) error {
+	endpoint := "/domains/"
+	pl := domainObject{Name: domain}
+	byt, _ := json.Marshal(pl)
+	var resp []byte
+	var err error
+	if resp, err = c.post(endpoint, "POST", byt); err != nil {
+		return fmt.Errorf("Error create domain deSEC: %v", err)
+	}
+	dm := domainObject{}
+	err = json.Unmarshal(resp, &dm)
+	if err != nil {
+		return err
+	}
+	printer.Printf("If you want to use DNSSec please add the DS record at your registrar using one of the keys:\n")
+	printer.Printf("%+q", dm.Keys)
+	return nil
+}
+
+//upsertRR will create or override the RRSet with the provided resource record.
+func (c *api) upsertRR(rr []resourceRecord, domain string) error {
+	endpoint := fmt.Sprintf("/domains/%s/rrsets/", domain)
+	byt, _ := json.Marshal(rr)
+	if _, err := c.post(endpoint, "PATCH", byt); err != nil {
+		return fmt.Errorf("Error create rrset deSEC: %v", err)
+	}
+	return nil
+}
+
+func (c *api) deleteRR(domain, shortname, t string) error {
+	endpoint := fmt.Sprintf("/domains/%s/rrsets/%s/%s/", domain, shortname, t)
+	if _, err := c.get(endpoint, "DELETE"); err != nil {
+		return fmt.Errorf("Error delete rrset deSEC: %v", err)
+	}
+	return nil
+}
+
+func (c *api) get(endpoint, method string) ([]byte, error) {
+	retrycnt := 0
+retry:
+	client := &http.Client{}
+	req, _ := http.NewRequest(method, "https://desec.io/api/v1"+endpoint, nil)
+	q := req.URL.Query()
+	req.Header.Add("Authorization", fmt.Sprintf("Token %s", c.creds.token))
+
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	bodyString, _ := ioutil.ReadAll(resp.Body)
+	// Got error from API ?
+	if resp.StatusCode > 299 {
+		if resp.StatusCode == 429 && retrycnt < 5 {
+			retrycnt++
+			time.Sleep(500 * time.Millisecond)
+			goto retry
```

Comment:
Handling retries that that is awesome. I need to copy that for digitalocean, which very much needs it.

Reply:
You can also look at the `Retry-After:` response header, which has the number of seconds you'll have to wait.

Reply:
Thanks.